cmd/adapter/adapter.go

```diff
@@ -0,0 +1,430 @@
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (ad *Adapter) Close() error {
+	var err error
+	if !utils.WaitWaitGroup(&ad.waitWg, time.Minute) {
+		err = errors.New("close timeout")
+	}
+	ad.logger.Info("Closed, err=", err)
+	return nil
+}
+
+//===================== adapter.api =====================
+
+func (ad *Adapter) runServeAPI(ctx context.Context) {
+	ad.logger.Info("Running serve API on ", ad.cfg.Gravity.ApiListenAddr)
+	ad.waitWg.Add(1)
+	go func() {
+		srv := ad.serveAPI(ctx)
+		select {
+		case <-ctx.Done():
+			sctx, _ := context.WithTimeout(context.Background(),
+				time.Second*time.Duration(5))
+			_ = srv.Shutdown(sctx)
+		}
+		ad.logger.Warn("Serving API stopped.")
+		ad.waitWg.Done()
+	}()
+}
+
+func (ad *Adapter) serveAPI(ctx context.Context) *http.Server {
+	router := httprouter.New()
+	router.GET("/v1/log", ad.makeHandlerWithCtx(ctx, ad.logHandler))
+	router.GET("/v1/download", ad.makeHandlerWithCtx(ctx, ad.downloadHandler))
+
+	srv := &http.Server{Addr: ad.cfg.Gravity.ApiListenAddr, Handler: router}
+	go func() {
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			ad.logger.Fatal("serveAPI(): %s", err)
+		}
+	}()
+	return srv
+}
+
+func (ad *Adapter) logHandler(ctx context.Context, rw http.ResponseWriter, rq *http.Request, p httprouter.Params) error {
+	ad.logger.Info("log(): Request=", rq)
+	qQuery := strings.TrimSpace(rq.URL.Query().Get("query"))
+	qLimit := strings.TrimSpace(rq.URL.Query().Get("limit"))
+
+	var (
+		err error
+	)
+
+	limit := lgLimitLines
+	if qLimit != "" {
+		limit, err = strconv.Atoi(qLimit)
+		if err != nil || limit < 0 {
+			limit, err = lgLimitLines, nil
+		}
+	}
+
+	qr := &api.QueryRequest{
+		Query: buildLql(qQuery, ad.cfg.Logrange.Partition, limit, -limit),
+		Pos:   "tail", Offset: -limit, Limit: limit,
+	}
+
+	ad.logger.Info("log(): Query=", qr.Query)
+	rs := &api.QueryResult{}
+	err = ad.lCli.Query(ctx, qr, rs)
+	if err == nil {
+		grLogEntries, err := ad.toGrLogEntries(rs.Events)
+		if err == nil {
+			bb, err := json.Marshal(grLogEntries)
+			if err == nil {
+				_, err = rw.Write(bb)
+			}
+		}
+	}
+
+	if err != nil {
+		ad.logger.Error("log(): Err=", err)
+		rw.WriteHeader(http.StatusInternalServerError)
+		_, _ = rw.Write([]byte("oops, something went wrong!\n"))
+	}
+
+	return err
+}
+
+func (ad *Adapter) toGrLogEntries(evs []*api.LogEvent) ([]string, error) {
+	gle := make([]string, 0, len(evs))
+	le := &grLogEntry{Type: "data"}
+	for _, e := range evs {
+		le.Payload = e.Message
+		bb, err := json.Marshal(le)
+		if err != nil {
+			return nil, err
+		}
+		gle = append(gle, string(bb))
+	}
+	return gle, nil
+}
+
+func (ad *Adapter) downloadHandler(ctx context.Context,
+	rw http.ResponseWriter, rq *http.Request, p httprouter.Params) error {
+
+	gzWriter := gzip.NewWriter(rw)
+	defer gzWriter.Close()
+	tarWriter := tar.NewWriter(gzWriter)
+	defer tarWriter.Close()
+
+	lql := buildLql("", ad.cfg.Logrange.Partition, dlLimitLines, -dlLimitLines)
```

Comment:
This function is pretty hard to follow - consider documenting its behavior and moving parts to auxiliary functions.

Reply:
applied (not yet merged to this branch)